Support multi-dim reduce op

barrh opened the pull request:
Prior to the patch, networks that apply reduce operation on
multiple dimensions have failed on ONNX JIT pass during
graph optimization.
e.g. torch.mean(tensor, [dim1, dim2])

This patch adds support for iterable dim argument.

Resolves issue #20942

ifedan commented on torch/onnx/symbolic.py:
```diff
@@ -80,7 +80,13 @@ def _parse_arg(value, desc):
+            try:
+                return [int(v) for v in tval]
+            except TypeError:
+                if tval.dim():
```
What is the purpose of calling dim() and raise the exceptions? Are you trying to figured out that tval is Tensor?

barrh replied:
Actually, I'm kinda assuming it's a tensor at that point. Trying to iterate over zero-dimension tensor throws a TypeError, so if the exception was raised due to that edge-case, I silence it.

barrh commented:
@pytorchbot rebase this please

barrh commented:
@yinghai would you mind approving this change?

houseroad commented on torch/onnx/symbolic.py:
```diff
@@ -80,7 +80,13 @@ def _parse_arg(value, desc):
+                if tval.dim():
+                    raise
+                # scalar tensors are not iterable
+                return [int(tval)]
```
Hi @barrh, this is not the right way to handle different type of inputs. You can use instance() function or _is_value + lengths.type().isSubtypeOf(torch._C.TensorType.get()). There are couples of examples in symbolic.py. Also, please add an end to end test case to test_pytorch_onnx_caffe2.py. We don't encourage adding more test cases to test_operators.py if unnecessary.

barrh replied:
@houseroad  I didn't understand your comment regarding the handling.
To allow this handling scale beyond 1 element, I iterate the tensor, but catch exception for scalar tensors (e.g. `torch.tensor(1)`. The TypeError is raised due to the traversing on non-iterable, not the element's type.

houseroad commented on torch/onnx/symbolic.py:
```diff
@@ -1195,7 +1202,7 @@ def feature_dropout(g, input, p, train):
 
 
-@parse_args('v', 't', 'i', 'i')
+@parse_args('v', 't', 'is', 'i')
```
Why change this?

barrh replied:
Similarly to mean, [torch.norm](https://pytorch.org/docs/stable/torch.html#torch.norm) accepts tuple types.
Before this change, `torch.norm(d, dim=(1,2))` would have failed.

houseroad replied:
if so, can we also add test cases?

barrh replied:
Similarly to mean, [torch.norm](https://pytorch.org/docs/stable/torch.html#torch.norm) accepts tuple types.
Before this change, torch.norm(d, dim=(1,2)) would have failed.

barrh replied:
added two tests.

houseroad reviewed (changes requested):
Please address the inline comments.

barrh committed:
Support multi-dim reduce ONNX ops

This patch adds support for iterable dim argument.
Prior to the patch, networks that apply reduce operation on
multiple dimensions have failed on ONNX JIT pass during
graph optimization.
e.g. torch.mean(tensor, [dim1, dim2])
e.g. torch.norm(tensor, dim=(dim1, dim2))

Resolves issue #20942

barrh commented:
@houseroad Do you have any idea how can I resolve the CI issues?
Not sure why I get the segfault, but it's pretty consistent, so it has to be introduced by my patch.
